### PR TITLE
rsx: Maintenance fixes [4]

### PR DESCRIPTION
**Pull request by kd-11**
- Do not enable host fp16 by default. Instead, move to GPU tab as a shader precision setting. More options will be moved under this switch in future. Native fp16 is actually too precise, compared to what RSX expects. That is what is causing the minor glitches when it is enabled. However, it generates much better shaders than the emulated fp16 version so I have left it in for users with weaker GPUs.
- Implement RSX-compliant polygon offset for the major GPU vendors. Polygon offset unfortunately is not portable, and different GPUs and drivers will have different calibration constants that need to be calculated and scaled to match RSX. I have covered AMD and NVIDIA on both windows and linux. Intel has not been tested, but should work just fine under vulkan. Intel + OpenGL is unsupported.

Fixes https://github.com/RPCS3/rpcs3/issues/11678
Fixes https://github.com/RPCS3/rpcs3/issues/11350
Fixes https://github.com/RPCS3/rpcs3/issues/11959

### CODE REVIEW

File: rpcs3/rpcs3qt/settings_dialog.ui [hunk 1]
```diff
@@ -43,8 +43,8 @@
        <rect>
         <x>0</x>
         <y>0</y>
-        <width>838</width>
-        <height>641</height>
+        <width>821</width>
+        <height>678</height>
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.ui:47]: Unnecessary changes

File: rpcs3/rpcs3qt/settings_dialog.ui [hunk 2]
```diff
@@ -54,7 +54,7 @@
        </sizepolicy>
       </property>
       <property name="currentIndex">
-       <number>0</number>
+       <number>1</number>
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.ui:57]: Unnecessary change

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 3]
```diff
@@ -546,6 +546,9 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
 		}
 	}
 
+	m_emu_settings->EnhanceComboBox(ui->shaderPrecision, emu_settings_type::ShaderPrecisionQuality);
+	SubscribeTooltip(ui->shaderPrecision, tooltips.settings.shader_precision);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.cpp:550]:
```suggestion
	SubscribeTooltip(ui->gbShaderPrecision, tooltips.settings.shader_precision);
```

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 4]
```diff
@@ -546,6 +546,9 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
 		}
 	}
 
+	m_emu_settings->EnhanceComboBox(ui->shaderPrecision, emu_settings_type::ShaderPrecisionQuality);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.cpp:549]: Lol there's literally comment for the combo box section below 🤣

File: rpcs3/rpcs3qt/emu_settings_type.h [hunk 5]
```diff
@@ -246,14 +246,14 @@ inline static const QMap<emu_settings_type, cfg_location> settings_location =
...
 	{ emu_settings_type::DisableFIFOReordering,      { "Video", "Disable FIFO Reordering"}},
-	{ emu_settings_type::DisableNativefloat16,       { "Video", "Disable native float16 support"}},
 	{ emu_settings_type::Enable3D,                   { "Video", "Enable 3D"}},
 	{ emu_settings_type::StrictTextureFlushing,      { "Video", "Strict Texture Flushing"}},
 	{ emu_settings_type::ForceCPUBlitEmulation,      { "Video", "Force CPU Blit"}},
 	{ emu_settings_type::DisableOnDiskShaderCache,   { "Video", "Disable On-Disk Shader Cache"}},
 	{ emu_settings_type::DisableVulkanMemAllocator,  { "Video", "Disable Vulkan Memory Allocator"}},
 	{ emu_settings_type::ShaderMode,                 { "Video", "Shader Mode"}},
 	{ emu_settings_type::ShaderCompilerNumThreads,   { "Video", "Shader Compiler Threads"}},
+	{ emu_settings_type::ShaderPrecisionQuality,     { "Video", "Shader Precision"}},
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/emu_settings_type.h:256]: Missing entry for the option in translation file

File: rpcs3/rpcs3qt/settings_dialog.ui (hunk 1 above)
Comment (kd-11) [reply rpcs3/rpcs3qt/settings_dialog.ui:47]: Blame QT creator.

File: rpcs3/rpcs3qt/emu_settings_type.h (hunk 5 above)
Comment (kd-11) [reply rpcs3/rpcs3qt/emu_settings_type.h:256]: We have a translation file? :laughing:

File: rpcs3/rpcs3qt/settings_dialog.cpp (hunk 4 above)
Comment (kd-11) [reply rpcs3/rpcs3qt/settings_dialog.cpp:549]: There's also one above.

File: rpcs3/rpcs3qt/settings_dialog.ui (hunk 2 above)
Comment (kd-11) [reply rpcs3/rpcs3qt/settings_dialog.ui:57]: What does this one mean? Every single time I open the file with QT creator it changes it to a 1.

Comment (Megamouse) [reply rpcs3/rpcs3qt/settings_dialog.ui]: It's the tab index you had open in the designer

File: rpcs3/Emu/system_config_types.cpp [hunk 6]
```diff
@@ -596,3 +596,18 @@ void fmt_class_string<thread_scheduler_mode>::format(std::string& out, u64 arg)
...
+
+template <>
+void fmt_class_string<gpu_preset_level>::format(std::string& out, u64 arg)
+{
+	format_enum(out, arg, [](gpu_preset_level value)
+	{
+		switch (value)
+		{
+		case gpu_preset_level::high: return "High";
+		case gpu_preset_level::low: return "Low";
```
Comment (Megamouse) [inline rpcs3/Emu/system_config_types.cpp:608]: what about _auto ???

File: rpcs3/rpcs3qt/settings_dialog.ui (hunk 2 above)
Comment (kd-11) [reply rpcs3/rpcs3qt/settings_dialog.ui]: Ah, makes sense. GPU is tab 1 so every time I make a change I have to switch to CPU tab in the designer before I save. What a wonderful design.

File: rpcs3/rpcs3qt/settings_dialog.cpp (hunk 4 above)
Comment (Megamouse) [reply rpcs3/rpcs3qt/settings_dialog.cpp:549]: well. you can remove the comment below then

File: rpcs3/rpcs3qt/emu_settings_type.h (hunk 5 above)
Comment (kd-11) [reply rpcs3/rpcs3qt/emu_settings_type.h:256]: I cannot find any references to a translation file for the other options, hint required.

Comment (Megamouse) [reply rpcs3/rpcs3qt/emu_settings_type.h:256]: emu_settings::GetLocalizedSetting

File: rpcs3/Emu/system_config_types.cpp (hunk 6 above)
Comment (kd-11) [reply rpcs3/Emu/system_config_types.cpp:609]: Added.

File: rpcs3/rpcs3qt/settings_dialog.cpp (hunk 4 above)
Comment (kd-11) [reply rpcs3/rpcs3qt/settings_dialog.cpp:549]: Removed.

**Comment by Linear524** [comment]
This PR fixes flickering shadows in God of War Ascension for my rigs (I use XeonV4 CPUs with TSX on @ GTX 1080)   ! :)
Thank yo KD-11 ! :)

**Comment by marcin-przywoski** [comment]
I think this commit has increased performance a lot in older R&C games on my 5950X. The 3rd planet in R&C2 (the space station) does not have 51 fps at the beginning but very close to locked 60 now. I think I'm seeing the same improvements in ToD

**Comment by Jorforit** [comment]
I don't understand the solution, it says to disable native fp16/ implement RSX-compliant polygon. what does it mean?
I don't know why but i also can't find the box to disable this/ don't know what an RSX-compliant polygon means.
Can someone help?

**Comment by Darkhost1999** [comment]
You don't need to do anything just update RPCS3 to reap the reward
